Rewrite: Applied DRY principle, better code structure & maintainability and new features

Techassi opened the pull request:
This PR introduces a complete rewrite of the tool. Most notably the following changes were made:

## CLI
Moving CLI related code into the `cmd` directory. This includes the command and prompts definitions

## Logic
Moving logic into the `internal` directory. All the logic resides in this folder and is split up by logical chunks:

- `config`: Reading, modifying, validating and writing the JSON config file
- `constants`: Constant values used across the tool
- `fixes`: Some prompt related fixes (see #8)
- `git`: Everything git command / config related
- `models`: Model definition, `User` and `Scope`
- `utils`: Small utility function used across the tool

## Breaking changes
The following breaking change was introduced: The ability to use `gitsu <alias>` to immediatly set a user profile (see #10 and #5) is now available via the `gitsu init <alias>` command.

## New features
gitsu now supports commands like:

```
gitsu
Use the arrow keys to navigate: ↓ ↑ → ← 
Select action
  ▶ Select user
    Add new user
    Delete user
    Modify user
```
---

```
gitsu select / gitsu s
```
With this command the user can immediatly proceed to select an existing user profile. Same as using `gitsu` and selecting __Select user__

---

```
gitsu add / gitsu a
```
With this command the user can immediatly proceed to add a new user profile. Same as using `gitsu` and selecting __Add new user__

---

```
gitsu delete / gitsu d
```
With this command the user can immediatly proceed to delete an existing user profile. Same as using `gitsu` and selecting __Delete user__

---

```
gitsu modify / gitsu m
```
With this command the user can immediatly proceed to modify an existing user profile. Same as using `gitsu` and selecting __Modify user__

---

```
gitsu reset / gitsu r
The following 2 user profile(s) will be deleted 
  [foo]  Bar <baz@example.com>
  [lazy] Dog <dog@example.com>

Use the arrow keys to navigate: ↓ ↑ → ← 
Delete above profiles?
  ▶ Yes
    No
```
With this command the user can reset / delete __ALL__ existing user profiles. This closes #4 

### Special alias 'default'
With the special alias `default` the user can specify a default user profile which gets set when executing `gitsu init` without supplying an alias (see __Breaking changes__).

### New fields in User model
Two new fields were added to the `User` model:

- `AddedAt`: Timestamp when the profile was added
- `ModifiedAt`: Timestamp when the profile was modified

Currently these fields aren't used, but will allow for future improvements like history for example.

Let me know what you think about this PR and if anything should be adjusted!

Techassi committed:
Remove old code

Techassi committed:
Update dependencies

Techassi committed:
Add CLI command definitions

Techassi committed:
Add prompt helpers

Techassi committed:
Add constants

Techassi committed:
Add bell skipper fix

Techassi committed:
Add core features

Techassi committed:
Remove docs / unused code

Techassi committed:
Add reset command

Techassi committed:
Mutiple changes: default profile, custom selection prompt

Add support for a default profile via the 'default' alias.

Add custom selection template with better styling.

Add GPG support and check if the user is within a git worktree.

Techassi committed:
Update dependencies

Techassi committed:
Add 'AddedAt' timestamp

Techassi committed:
Merge remote-tracking branch 'upstream/main' into rewrite

matsuyoshi30 commented:
@Techassi Thanks! I'll take a look later :)

matsuyoshi30 commented on cmd/add.go:
```diff
@@ -0,0 +1,67 @@
+			}
+
+			user := models.NewUser(name, email, alias, keyID)
+			cfg, err := config.Read()
```
Those who use it for the first time cannot add a git user because `config.Read()` returns an error (config file doesn't exist).

I think it has to check config file existence before `config.Read()` and create empty `*Config` if it doesn't exist.

Techassi replied:
This is now fixed by commit 01df57b. Let me know if everything works as expected!

matsuyoshi30 commented:
I think it's great PR overall. Please check my comment about `cmd/add.go`. I'm ready to merge this! @Techassi

Techassi committed:
Fix 'Config file does not exist' error when using the tool for the first time

This is done with the new function CreateEmptyConfigIfNeeded which creates a new empty config if it does not exist yet.

Threpio commented:
👏 👏 👏 👏

matsuyoshi30 commented:
@Techassi Thanks for your contribution again!

Techassi commented:
The workflow of this tool changed quite a bit over the last time. We should consider adjusting the GIF in the README to give a better overview of the (changed) features. I can also open an issue to track this.
